Criação do documento de padrões de projetos emergentes (#94)
UnBArqDsw2021-2/2021.2_G4_CadernetaDeCampoDigital_docs: docs/padroes_projetos/padroes_emergentes.md

# Padrões de Projetos Emergentes

## 1. Versionamento

| Versão | Data       | Descrição                                             | Autor(es)                     |
| ------ | ---------- | ----------------------------------------------------- | ----------------------------- |
| 1.0    | 24/02/2022 | Criação do documento e introdução                     | Vitor Lamego |
| 1.1    | 24/02/2022 | Adição dos Padrões Emergentes                     | Vitor Lamego |
| 1.2    | 24/02/2022 | Adição dos Padrões Rest e MVT                     | Denniel William |

## 2. Introdução

> <p align="justify" style="text-indent: 20px">Padrões de projeto são formatos de soluções que são utilizadas de forma recorrente dentro do projeto. Esses padrões são comumente utilizados em projetos com o paradigma Orientado a Objeto, sendo bastante, tranformando esses padrões em artifícios de comunicação e solução dentro das equipes.[1]</p>

Reviewer: Não entendi a frase "... Orientado a Objeto, sendo bastante, tranformando ...", ficou estranho esse sendo bastante junto com o transformando

> <p align="justify" style="text-indent: 20px">Os padrões nomeados como emergentes são derivados da necessidade de algum fator específico para utilizar determinada solução. Portanto, os padrões emergentes são definidos quando possuímos um determinado comportamento inerente ao padrão, contudo não possui a estrutura que o descreve, o que é muito comum em situações como: falta de conhecimento por parte do desenvolvedor sobre os padrões de projetos, sucessivas mudanças em outras classes que então é refletida em uma determinada parte do código, falta de refatoração, característica do problema e, então, não necessita implementar o padrão.[1]</p>

Reviewer: ```suggestion
<p align="justify" style="text-indent: 20px">Os padrões nomeados como emergentes são derivados da necessidade de algum fator específico para utilizar determinada solução. Portanto, os padrões emergentes são definidos quando possuímos um determinado comportamento inerente ao padrão, contudo não possuimos a estrutura que o descreve, o que é muito comum em situações como: falta de conhecimento por parte do desenvolvedor sobre os padrões de projetos, sucessivas mudanças em outras classes que então são refletida em uma determinada parte do código, falta de refatoração, característica do problema e, então, não necessita implementar o padrão.[1]</p>
```
Além disso, a frase "... e, então, não necessita implementar o padrão.[1]" ficou sem sentido para mim.

## 3. Exemplos de Padrões Emergentes

### 3.1 Singleton

> <p align="justify" style="text-indent: 20px">Existindo uma determinada classe A, se esta for um Singleton, em qualquer momento de execução essa classe pode ter apenas zero ou uma instância dessa classe. Portando todos os caminhos possíveis do programa, se necessário, devem utilizar essa única instância e não criar uma nova. É muito comum esse comportamento ser repetido por classes dentro de um projeot, mas não respeitando a sua estrutrua para ser considerada um padrão de projeto, nessas ocasiões o Singleton acaba sendo um padrão que emerge naquela parte específica do código.</p>

Reviewer: Se formos seguir o definido pela professora, Singleton é considerado um GOF Criacional. Rever esse tópico.

### 3.2 Injeção de Dependência

<p align="justify" style="text-indent: 20px">Injeção de dependência é um padrão utilizado para dimimuir o acoplamento de determinada parte do código, que costuma acontecer quando uma classe depende da instância de uma outra classe para a sua criação. Dessa forma a interface é a responsável por injetar em cada classe as suas dependências declaradas. Nesta ocasião a injeção de dependência se relaciona com o padrão de "Inversão de controle", mas não pode ser considerado como um sinônimo, ou seja, a responsabilidade de informar a implementação utilizada deixa de ser da classe e passa a ser de quem está consumindo a classe.[2]</p>

> <p align="justify" style="text-indent: 20px">O baixo acoplamento é importante para facilitar a manutenibilidade do código, além de permitir a utilização de mocks para realizar testes unitários nas respecivas classes.</p>

Reviewer: O baixo acoplamento é gerado da injeção de dependência ? Se sim, explicar melhor porque ficou bem solto esse parágrafo.

### 3.3 Factory Method

> <p align="justify" style="text-indent: 20px">O padrão de Factory Method é utilizado quando temos um método m que é responsável por retornar um objeto do tipo T, que podem, a qualquer momento, serem instanciadas apenas a partir do método m. Um exemplo claro de quando temos o Factory Method como um padrão emergente é quando não existem subclasses que permitam a escolha de como será implementado as instâncias desse objeto T.[1]</p>

Reviewer: Factory Method assim como os Singletons são padrões de GOF's criacionais. Um dos principais é exatamente o factory method. Rever esse parágrafo.

### 3.3 Rest

> <p align="justify" style="text-indent: 20px">O padrão de Rest ou Restful, Representational State Transfer (Transferência de Estado Representacional), define a padronização de rotas, requisições e comunicações sem estado, também conhecico como protocolo HTTP, sendo assim o padrão rest define de utilizar desse protocolo sendo que a API que possui a inteligência de aplicar esse padrão ela tem termo atribuído como Restful.[1]</p>

Reviewer: O padrão não é REST ou RESTful. Ele utiliza a uma filosofica RESTful para aplicar o padrão REST (conforme explicação do @thiagohdaqw no dojo de django).

```suggestion
<p align="justify" style="text-indent: 20px">O padrão de REST ou RESTful, Representational State Transfer (Transferência de Estado Representacional), define a padronização de rotas, requisições e comunicações sem estado, também conhecico como protocolo HTTP. Sendo assim, o padrão REST define de utilizar desse protocolo sendo que a API que possui a inteligência de aplicar esse padrão ela tem termo atribuído como Restful.[1]</p>
```

Ficou sem sentido a frase "... o padrão REST define de utilizar desse protocolo sendo que a API que ...".

O número dessa referência está certo ? Todos vem da 1, sendo que existem 5 referência citadas ?

### 3.4 Padrão MVT (Model - View - Template)

> <p align="justify" style="text-indent: 20px">Um dos padrões também que serão seguidos, será o MVT, em que por ter sido de opção da equipe a utilização do framework Django, o padrão a ser utilizado será o MVT.[1]</p>

Reviewer: ```suggestion
<p align="justify" style="text-indent: 20px">Um dos padrões que também serão seguidos, será o MVT, em decorrência da utilização do framework Django, definido pela equipe para o backend</p>
```
Tentei dar uma melhorada nesse parágrafo, porque ele ficou bem repetitivo sem acrescentar muitas informações ao tópico.

> <p align="justify" style="text-indent: 20px">O padrão MVT se baseia na distribuição de Model, View e Template: A model atua como a interface de dados, ela é a estrutura lógica por trás do aplicativo e é representado pelo banco de dados do projeto; A view é usado para executar as regras de negócios, interação com o modelo para transporte de dados e renderização do template; O template é a camada de aprensetação de trata da parte de interface do usuário.</p>

Reviewer: ```suggestion
<p align="justify" style="text-indent: 20px">O padrão MVT se baseia na distribuição de Model, View e Template: A model atua como a interface de dados, ela é a estrutura lógica por trás do aplicativo e é representado pelo banco de dados do projeto; A view é usada para executar as regras de negócios, interação com o modelo para transporte de dados e renderização do template; O template é a camada de apresentação, e é responsável pela parte da interface do usuário.</p>
```
Esse texto tem alguma referência ? Porque se não, seria bom botar. Além disso, a gente não vai estar utilizando todo o MVT, a princípio, o frontend vai ser o responsável pela camada de "template".

## 4. Referências

> [1] JOB, Ricardo de Sousa. Uma abordagem para detecção de padrões emergentes. 2014. 92f. (Dissertação de Mestrado em Ciência da Computação) Programa de Pós-graduação em Ciência da Computação, Centro de Engenharia Elétrica e Informática, Universidade Federal de Campina Grande - Paraiba - Brasil, 2014.

> [2] **Injeção de Dependência**. O que é injeção de dependência ? Disponível em: <a href="https://medium.com/@eduardolanfredi/inje%C3%A7%C3%A3o-de-depend%C3%AAncia-ff0372a1672" >Injeção de Dependência</a>. Acesso em: 24 de fev. de 2022.

> [3] **Design de API Rest**. Design de API Rest. Disponível em: <a href="https://wssilva-willian.medium.com/design-de-api-rest-9807a5b16c9f" >Design de API Rest</a>. Acesso em: 25 de fev. de 2022.

> [4] **Django MVT**. Django MVT. Disponível em: <a href="https://www.javatpoint.com/django-mvt" >Django MVT</a>. Acesso em: 25 de fev. de 2022.

> > [5] **Entendendo o MTV do Django**. Entendendo o MTV do Django. Disponível em: <a href="https://www.treinaweb.com.br/blog/entendendo-o-mtv-do-django" >Entendendo o MVT do Django</a>. Acesso em: 25 de fev. de 2022.

Reviewer: Dessas 5 referências, só vi 2 serem utilizadas. Ou foram citadas referências não utilizadas, ou elas não foram referênciadas corretamente nos textos.

### 3.2 Rest

<p align="justify" style="text-indent: 20px">O padrão REST, Representational State Transfer (Transferência de Estado Representacional), define a padronização de rotas, requisições e comunicações sem estado, também conhecido como protocolo HTTP. Sendo assim, uma API que utiliza essa padronização e que possui inteligência para aplicá-la é denominada de RESTful. [3]</p>

<p align="justify" style="text-indent: 20px">Existem 4 princípios principais do REST: [5]</p>

- **Client-Server (Cliente-Servidor):** Sempre há um cliente e um servidor e esses dois sistemas precisam de limites para o funcionamento.[5]
- **Stateless (Sem estado):** Os servidores precisam ser capazes de processar as mensagens que recebem. Para fazer isso, cada solicitação que um servidor receber deve ter as informações necessárias para que o servidor funcione.[5]
- **Uniform Interface (Interface Uniforme):** O uso de terminologia e recursos semelhantes ajuda a padronizar as APIs. De acordo com o princípio, os seguintes verbos HTTP são usados: GET, PUT, POST,DELETE. Os recursos sempre se referem a URIs (identificador uniforme de recursos). As respostas HTTP sempre vêm com um status e um corpo.[5]

> - **Cacheable (Armazenável com cache):** Os clientes precisam ser capazes de armazenar em cache as representações. Devido à ausência de estado (toda representação sendo autodescritiva), isso é possível em uma API RESTful.[5]

Reviewer: Aqui como já foi citado o [5] no parágrafo de introdução, pode tirar a numeração de cada tópico

> <img src="../../../assets/padroes_projeto/rest.png" width="800"/>

Reviewer: Centralizar essa e a imagem abaixo
